distribution/pull_v2_windows.go

```diff
@@ -62,24 +63,27 @@ func (ld *v2LayerDescriptor) open(ctx context.Context) (distribution.ReadSeekClo
 	return rsc, err
 }
 
-func filterManifests(manifests []manifestlist.ManifestDescriptor, os string) []manifestlist.ManifestDescriptor {
-	osVersion := ""
-	if os == "windows" {
-		version := system.GetOSVersion()
-		osVersion = fmt.Sprintf("%d.%d.%d", version.MajorVersion, version.MinorVersion, version.Build)
-		logrus.Debugf("will prefer entries with version %s", osVersion)
-	}
+func filterManifests(manifests []manifestlist.ManifestDescriptor, p specs.Platform) []manifestlist.ManifestDescriptor {
+	version := system.GetOSVersion()
+	osVersion := fmt.Sprintf("%d.%d.%d", version.MajorVersion, version.MinorVersion, version.Build)
+	logrus.Debugf("will prefer Windows entries with version %s", osVersion)
 
 	var matches []manifestlist.ManifestDescriptor
+	foundWindowsMatch := false
 	for _, manifestDescriptor := range manifests {
-		if manifestDescriptor.Platform.Architecture == runtime.GOARCH && manifestDescriptor.Platform.OS == os {
+		if (manifestDescriptor.Platform.Architecture == runtime.GOARCH) &&
+			((p.OS != "" && manifestDescriptor.Platform.OS == p.OS) || // Explicit user request for an OS we know we support
+				(p.OS == "" && system.IsOSSupported(manifestDescriptor.Platform.OS))) { // No user requested OS, but one we can support
```

Comment:
With this change there can be multiple operating systems in the list, which `manifestsByVersion` does not currently handle properly. This means if Linux comes before Windows in the manifest list, then Linux will be selected unless an exact Windows build match is found. With Hyper-V containers, other versions of Windows should still be preferred over Linux, but that is not happening (17134 host, 16299 highest in list):

```
$ docker image pull hello-seattle
Using default tag: latest
latest: Pulling from library/hello-seattle
6d7e9a8e8043: Pull complete
Digest: sha256:93cdcb901826410633ef2a18052b6a1ffc72f14a3323ddc94023ff58f4171d49
Status: Downloaded newer image for hello-seattle:latest

$ (docker image inspect hello-seattle | ConvertFrom-Json).Os
linux
```

The following  change addresses this by always sorting Windows first:

```diff
@@ -103,7 +103,8 @@ func (mbv manifestsByVersion) Less(i, j int) bool {
        // TODO: Split version by parts and compare
        // TODO: Prefer versions which have a greater version number
        // Move compatible versions to the top, with no other ordering changes
-       return versionMatch(mbv.list[i].Platform.OSVersion, mbv.version) && !versionMatch(mbv.list[j].Platform.OSVersion, mbv.version)
+       return (strings.EqualFold("windows", mbv.list[i].Platform.OS) && !strings.EqualFold("windows", mbv.list[j].Platform.OS)) ||
+               (versionMatch(mbv.list[i].Platform.OSVersion, mbv.version) && !versionMatch(mbv.list[j].Platform.OSVersion, mbv.version))
 }
```

With this change, the expected images are pulled:

```
$ docker image pull hello-seattle
Using default tag: latest
latest: Pulling from library/hello-seattle
bce2fbc256ea: Already exists
83eec61707e8: Already exists
32e7fb1c173a: Pull complete
8d9eb6464441: Pull complete
Digest: sha256:93cdcb901826410633ef2a18052b6a1ffc72f14a3323ddc94023ff58f4171d49
Status: Downloaded newer image for hello-seattle:latest

$ (docker image inspect hello-seattle | ConvertFrom-Json).Os
windows

$ docker image pull --platform linux hello-seattle
Using default tag: latest
latest: Pulling from library/hello-seattle
6d7e9a8e8043: Pull complete
Digest: sha256:93cdcb901826410633ef2a18052b6a1ffc72f14a3323ddc94023ff58f4171d49
Status: Downloaded newer image for hello-seattle:latest

$ (docker image inspect hello-seattle | ConvertFrom-Json).Os
linux
```

Without Hyper-V, the image build must match the host. In that case, it's possible that none of the Windows images in the manifest list will run, but the Linux one may. However, it might still be acceptable to prefer Windows and show the error, informing the user that the manifest list contains Windows, but the wrong version. The user can then choose to pull the Linux image with the platform switch.

On a related note, Hyper-V does not support running builds that are _higher_ than the host. Those should be filtered out entirely, and never added to the list. I will follow up with a separate fix for that.

@jhowardmsft, please take a look.